code/controller/astar.cpp

```diff
@@ -0,0 +1,109 @@
+#include "astar.h"
+#include <cmath>
+#include <set>
+#include <queue>
+#include <iostream>
+
+using namespace std;
+
+Astar::Astar() {}
+
+double Astar::manhattanDist(Astar::Coord cur, Astar::Coord dest) {
+    return abs(dest.x - cur.x) + abs(dest.y - cur.y);
+}
+
+bool Astar::isValid(int x, int y, int maxRow, int maxCol) {
+    return (x >= 0) && (x < maxRow) && (y >= 0) && (y < maxCol);
+}
+
+vector<Astar::Coord> Astar::getNeighbours(Astar::Coord c, int maxRow, int maxCol) {
+    vector<Astar::Coord> neighbours;
+    int cur_x = c.x;
+    int cur_y = c.y;
+    if (isValid(cur_x-1, cur_y, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x-1, cur_y});
+    }
+    if (isValid(cur_x+1, cur_y, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x+1, cur_y});
+    }
+    if (isValid(cur_x, cur_y-1, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x, cur_y-1});
+    }
+    if (isValid(cur_x, cur_y+1, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x, cur_y+1});
+    }
+
+    return neighbours;
+}
+
+// For testing purposes
+Astar::grid Astar::makeTerrain() {
+    grid a = {{1,   1, -1, 1},
+              {1,   1, -1, 1},
+              {-1,  1, 1,  1},
+              {1,   1, -1, 1}};
+    return a;
+}
+
+void Astar::backtrack(Astar::Coord start, Astar::Coord dest, map<Astar::Coord, Astar::Coord> parent) {
+    Astar::Coord cur = dest;
+    while (cur != start) {
+        path.push_back(cur);
+        cur = parent[cur];
+    }
+    path.push_back(start);
+}
+
+void Astar::searchPath(Astar::grid terrain, Astar::Coord start, Astar::Coord dest) {
+    // idea for using maps to keep track of path information came from redblobgames.com
+    map<Astar::Coord, Astar::Coord> parent;
```

Comment:
technically if you used `unordered_map` and made a hash function for `Astar::Coord` you could design a perfect hash function for a given hash table size so you'd never get collisions, then it'd be way faster than using `map` which has an RB tree behind it

Reply:
actually that's probably more effort than it's worth